Wikipedia/Code/WMFZeroConfigurationManager.m

```diff
@@ -0,0 +1,166 @@
+        }
+    }
+}
+
+- (BOOL)isZeroRated {
+    BOOL isZeroRated;
+    @synchronized(self) {
+        isZeroRated = _isZeroRated;
+    }
+    return isZeroRated;
+}
+
+- (void)setWarnWhenLeaving:(BOOL)warnWhenLeaving {
+    [[NSUserDefaults wmf_userDefaults] setObject:[NSNumber numberWithBool:warnWhenLeaving]
+                                          forKey:WMFZeroWarnWhenLeaving];
+    [[NSUserDefaults wmf_userDefaults] synchronize];
+}
+
+- (BOOL)warnWhenLeaving {
+    return [[NSUserDefaults wmf_userDefaults] boolForKey:WMFZeroWarnWhenLeaving];
+}
+
+#pragma mark - Banner Updates
+
+/**
+ *   This method:
+ *
+ * - Fetches carrier specific strings placing them in self.zeroConfiguration object.
+ *
+ * - If the fetched zeroConfiguration has a nil "message" string, this means even though
+ *   there was a header, leading us to believe this network was Zero rated, its Zero
+ *   rating is not presently enabled. (It would be nice if the query fetching the
+ *   zeroConfiguration returned an "enabled" key/value, but it doesn't - it nils out the
+ *   values instead apparently.) So in the nil message case we set isZeroRated "NO".
+ */
+- (void)fetchZeroConfigurationAndSetIsZeroRatedIfNecessary {
+    
+    // Note: don't nil out self.zeroConfiguration in this method
+    // because if we do we can't show its exit message strings!
+
+    //TODO: ensure thread safety so we can do this work off the main thread...
+    dispatch_async(dispatch_get_main_queue(), ^{
+        @weakify(self);
+        AnyPromise *promise = [AnyPromise promiseWithValue:nil];
+        promise = [self fetchZeroConfiguration].then(^(WMFZeroConfiguration *zeroConfiguration) {
+            @strongify(self);
+            
+            // If the config is not enabled its "message" will be nil, so if we detect a nil message
+            // set the isZeroRated to NO before we post the WMFZeroRatingChanged notification.
+            if(zeroConfiguration.message == nil){
+                self.isZeroRated = NO;
+                // Reminder: don't nil out self.zeroConfiguration here or the carrier's exit message won't be available.
+            }else{
+                self.zeroConfiguration = zeroConfiguration;
+                self.isZeroRated = YES;
+            }
+            
+        }).catch(^(NSError* error){
+            @strongify(self);
+            self.isZeroRated = NO;
+        });
+    });
+}
+
+- (AnyPromise *)fetchZeroConfiguration {
+    [self.zeroConfigurationFetcher cancelAllFetches];
+    return [self.zeroConfigurationFetcher fetchZeroConfigurationForSiteURL:[[[MWKLanguageLinkController sharedInstance] appLanguage] siteURL]];
+}
+
+- (void)updateZeroRatingAndZeroConfigurationForResponseHeadersIfNecessary:(NSDictionary *)headers {
+    NSAssert(headers != nil, @"Expecting response headers.");
+    if(!headers){
+        return;
+    }
+
+    bool zeroEnabled = self.isZeroRated;
```

Comment:
should be `BOOL` - `bool` is standard a C type where `BOOL` is a typedef that Apple defines